apollo-router/src/query_planner/bridge_query_planner_pool.rs

```diff
@@ -0,0 +1,200 @@
+use crate::services::QueryPlannerResponse;
+use crate::spec::Schema;
+use crate::Configuration;
+
+static CHANNEL_SIZE: usize = 10_000;
+
+#[derive(Clone)]
+pub(crate) struct BridgeQueryPlannerPool {
+    planners: Vec<Arc<Planner<QueryPlanResult>>>,
+    sender: Sender<(
+        QueryPlannerRequest,
+        oneshot::Sender<Result<QueryPlannerResponse, QueryPlannerError>>,
+    )>,
+    schema: Arc<Schema>,
+    subgraph_schemas: Arc<HashMap<String, Arc<Valid<apollo_compiler::Schema>>>>,
+}
+
+impl BridgeQueryPlannerPool {
+    pub(crate) async fn new(
+        sdl: String,
+        configuration: Arc<Configuration>,
+        size: NonZeroUsize,
+    ) -> Result<Self, ServiceBuildError> {
+        Self::new_from_planners(Default::default(), sdl, configuration, size).await
+    }
+
+    pub(crate) async fn new_from_planners(
+        old_planners: Vec<Arc<Planner<QueryPlanResult>>>,
+        schema: String,
+        configuration: Arc<Configuration>,
+        size: NonZeroUsize,
+    ) -> Result<Self, ServiceBuildError> {
+        let mut join_set = JoinSet::new();
+
+        let (sender, receiver) = bounded::<(
+            QueryPlannerRequest,
+            oneshot::Sender<Result<QueryPlannerResponse, QueryPlannerError>>,
+        )>(CHANNEL_SIZE);
+
+        let mut old_planners_iterator = old_planners.into_iter();
+
+        (0..size.into()).for_each(|_| {
+            let sdl = schema.clone();
+            let configuration = configuration.clone();
+
+            if let Some(old_planner) = old_planners_iterator.next() {
+                join_set.spawn(async move {
+                    BridgeQueryPlanner::new_from_planner(old_planner, sdl, configuration).await
+                });
+            } else {
+                join_set.spawn(async move { BridgeQueryPlanner::new(sdl, configuration).await });
+            }
+        });
+
+        let mut bridge_query_planners = Vec::new();
+
+        while let Some(task_result) = join_set.join_next().await {
+            let bridge_query_planner =
+                task_result.map_err(|e| ServiceBuildError::ServiceError(Box::new(e)))??;
+            bridge_query_planners.push(bridge_query_planner);
+        }
+
+        let schema = bridge_query_planners
+            .first()
+            .expect("There should be at least 1 service in pool")
+            .schema();
+
+        let subgraph_schemas = bridge_query_planners
+            .first()
+            .expect("There should be at least 1 service in pool")
+            .subgraph_schemas();
+
+        let planners = bridge_query_planners
+            .iter()
+            .map(|p| p.planner().clone())
+            .collect();
+
+        for (worker_id, mut planner) in bridge_query_planners.into_iter().enumerate() {
+            let receiver = receiver.clone();
+
+            tokio::spawn(async move {
+                while let Ok((request, res_sender)) = receiver.recv().await {
+                    let start = Instant::now();
```

Comment:
Should this be at line 116? Otherwise the metric includes waiting for the planner to become ready.

Reply:
Fixed in [1f7e478](https://github.com/apollographql/router/pull/4897/commits/1f7e478781f24f12f064ca669cddd8688b82db6d)